test/e2e/storage/pd.go

```diff
@@ -59,467 +60,338 @@ var _ = SIGDescribe("Pod Disks", func() {
-			Expect(len(nodes.Items)).To(Equal(initialGroupSize), "Requires node count to return to initial group size.")
-		}()
-
-		By("submitting host0Pod to kubernetes")
-		_, err = podClient.Create(host0Pod)
-		framework.ExpectNoError(err, fmt.Sprintf("Failed to create host0pod: %v", err))
-
-		framework.ExpectNoError(f.WaitForPodRunningSlow(host0Pod.Name))
-
-		testFile := "/testpd1/tracker"
-		testFileContents := fmt.Sprintf("%v", mathrand.Int())
-
-		framework.ExpectNoError(f.WriteFileViaContainer(host0Pod.Name, containerName, testFile, testFileContents))
-		framework.Logf("Wrote value: %v", testFileContents)
-
-		// Verify that disk shows up in node 0's volumeInUse list
-		framework.ExpectNoError(waitForPDInVolumesInUse(nodeClient, diskName, host0Name, nodeStatusTimeout, true /* should exist*/))
-
-		gceCloud, err := framework.GetGCECloud()
-		framework.ExpectNoError(err, fmt.Sprintf("Unable to create gcloud client err=%v", err))
-
-		output, err := gceCloud.ListInstanceNames(framework.TestContext.CloudConfig.ProjectID, framework.TestContext.CloudConfig.Zone)
-		framework.ExpectNoError(err, fmt.Sprintf("Unable to get list of node instances err=%v output=%s", err, output))
-		Expect(true, strings.Contains(string(output), string(host0Name)))
-
-		By("deleting host0")
-		resp, err := gceCloud.DeleteInstance(framework.TestContext.CloudConfig.ProjectID, framework.TestContext.CloudConfig.Zone, string(host0Name))
-		framework.ExpectNoError(err, fmt.Sprintf("Failed to delete host0pod: err=%v response=%#v", err, resp))
-
-		output, err = gceCloud.ListInstanceNames(framework.TestContext.CloudConfig.ProjectID, framework.TestContext.CloudConfig.Zone)
-		framework.ExpectNoError(err, fmt.Sprintf("Unable to get list of node instances err=%v output=%s", err, output))
-		Expect(false, strings.Contains(string(output), string(host0Name)))
+	Context("detach from a disrupted node [Slow] [Disruptive]", func() {
+		const (
+			deleteNode    = 1 // delete physical node
+			deleteNodeObj = 2 // delete node's api object only
+		)
+		type testT struct {
+			descr  string // It description
+			nodeOp int    // disruptive operation performed on target node
+		}
+		tests := []testT{
+			{
+				descr:  "node is deleted",
+				nodeOp: deleteNode,
+			},
+			{
+				descr:  "node's API object is deleted",
+				nodeOp: deleteNodeObj,
+			},
+		}
 
-		// The disk should be detached from host0 on it's deletion
-		By("Waiting for pd to detach from host0")
-		waitForPDDetach(diskName, host0Name)
-		framework.ExpectNoError(framework.WaitForGroupSize(framework.TestContext.CloudConfig.NodeInstanceGroup, int32(initialGroupSize)), "Unable to get back the cluster to inital size")
-		return
-	})
+		for _, t := range tests {
+			nodeOp := t.nodeOp
+			It(fmt.Sprintf("when %s", t.descr), func() {
+				framework.SkipUnlessProviderIs("gce")
+
+				initialGroupSize, err := framework.GroupSize(framework.TestContext.CloudConfig.NodeInstanceGroup)
```

Comment:
You may want to use [`framework.GetReadySchedulableNodesOrDie()`](https://github.com/kubernetes/kubernetes/blob/5af069b727debe8b2ea075d0514ad6099733be90/test/e2e/framework/util.go#L2458).  It uses the kube api to get a list of reachable, healthy nodes.  [`framework.GroupSize()`](https://github.com/kubernetes/kubernetes/blob/5af069b727debe8b2ea075d0514ad6099733be90/test/e2e/framework/size.go#L85) instead uses the `gcloud` or `aws` cli to get the number of platform instances (VMs), regardless of whether the kubelet is running.  So you could end up with an miscount of reachable nodes.

Also, using the cli is inherently slow.  I know gcloud can take whole seconds to complete an op.

Reply:
That said, you'll need to check the `len()` of the `nodeList.Items` that ` framework.GetReadySchedulableNodesOrDie()` returns.  There is no error if no nodes are schedulable.

Reply:
ack and thanks for the suggestion!